### PR TITLE
CLOUDP-87338: Implement PrivateEndpoint support

### PR DESCRIPTION
**Pull request by vasilevp**
Steps to create a Private Endpoint:
1. Add PE to the plan
2. Create Private Endpoint Service in Atlas
3. Create Private Endpoint in Azure
3.1 Disable policies for target subnet
3.2 Create Private Endpoint
4. Create Private Endpoint Connection in Atlas

FYI: 
- For some reason, CF refuses to build AOSB after including the Azure SDK - this is most likely due to a build timeout, and needs fixing
- This works in the simplest scenario (1 endpoint per cluster, no collisions with existing resources) after working around the above issue
- Azure side of things is non-intuitive, takes a lot of steps to do one thing, and thus can be unstable at the moment
- Proper Azure auth is not yet implemented

### CODE REVIEW

File: pkg/broker/privateendpoint/privateendpoint.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	SubnetName         string `json:"subnetName,omitempty"`
+	EndpointName       string `json:"endpointName,omitempty"`
+}
+
+// TODO: this is a hack to get Azure working ASAP
+// replace with proper auth (see below)
+type tokenProvider struct{}
+
+func (*tokenProvider) OAuthToken() string {
+	return os.Getenv("AZURE_BEARER_TOKEN")
```
Comment (vasilevp) [inline pkg/broker/privateendpoint/privateendpoint.go:55]: To get the token run `az account get-access-token`
The token expires in 1 hour (barely enough for testing).

File: samples/plans/sample_basic.yml.tpl [hunk 2]
```diff
@@ -54,3 +54,13 @@ ipAccessLists:
   comment: "everything"
 - ipAddress: "128.0.0.0/1"
   comment: "everything"
+
+#privateEndpoints:
```
Comment (vasilevp) [inline samples/plans/sample_basic.yml.tpl:58]: Just an example - remove when ready (or create a separate plan for this)

File: pkg/broker/instance_operations.go [hunk 3]
```diff
@@ -188,9 +189,104 @@ func (b *Broker) createOrUpdateResources(ctx context.Context, client *mongodbatl
...
+			}
+
+			for _, e := range service.Endpoints {
+				future, err := privateendpoint.Create(ctx, e, atlasService)
+				if err != nil {
+					return false, errors.Wrap(err, "cannot create Private Endpoint in Azure")
+				}
+
+				pe, err := future()
+				if err != nil {
```
Comment (vasilevp) [inline pkg/broker/instance_operations.go:243]: As per the docs, error means the `future` is not ready yet

File: pkg/broker/privateendpoint/privateendpoint.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		}
+	} else {
+		// TODO: handle existing endpoint properly
+		return func() (network.PrivateEndpoint, error) {
+			return ep, nil
+		}, nil
+	}
+
+	future, err := peClient.CreateOrUpdate(ctx, e.ResourceGroup, e.EndpointName, network.PrivateEndpoint{
+		Location: to.StringPtr("westus"), // TODO: add another field to plan? or deduce from Atlas region name?
```
Comment (vasilevp) [inline pkg/broker/privateendpoint/privateendpoint.go:104]: `westus` is what I got during testing - YMMV

File: pkg/broker/instance_operations.go [hunk 5]
```diff
@@ -462,6 +590,20 @@ func (b Broker) LastOperation(ctx context.Context, instanceID string, details do
...
 		}
 
+		retry := false
+		logger.Debugw("Create resources", "plan", p)
+		retry, err = b.postCreateResources(ctx, client, p)
+		if err != nil {
+			break
+		}
+
+		if retry {
```
Comment (antonlisovenko) [inline pkg/broker/instance_operations.go:600]: [q] am I right that the `retry` variable affects results in some retry of the logic in the code? I don't see where this variable is used though - could you point please?

File: pkg/broker/instance_operations.go [hunk 6]
```diff
@@ -188,9 +193,104 @@ func (b *Broker) createOrUpdateResources(ctx context.Context, client *mongodbatl
...
+			}
+
+			continue
+		}
+
+		if existing.PrivateEndpointIPAddress == addr {
+			continue
+		}
+
+		_, err = client.PrivateEndpoints.DeleteOnePrivateEndpoint(ctx, dp.Project.ID, endpoint.Provider, endpoint.ID, *pe.ID)
```
Comment (antonlisovenko) [inline pkg/broker/instance_operations.go:283]: [q] why are we deleting the private endpoint?

**Review by antonlisovenko — COMMENTED** [review]
This looks great to me, asked a couple of questions.

File: pkg/broker/instance_operations.go (hunk 5 above)
Comment (fabritsius) [reply pkg/broker/instance_operations.go:622]: > // TODO: this retry logic is clunky, come up with something better?
The way I see it `postCreateResources` method returns `retry` as sort of a flag. Which indicates if the same request should be repeated. 

```go
if retry {
	resp.State = domain.InProgress
	resp.Description = "resources are being created"

	break
}
```
Here we break if this flag is `true`.

This solution seems temporary but I haven't yet started thinking of a better alternative.

File: pkg/broker/instance_operations.go (hunk 6 above)
Comment (fabritsius) [reply pkg/broker/instance_operations.go:305]: Not quite sure, but my best guess is that it prevents half-configured endpoints from interrupting the next try.

If everything goes fine we return/continue before this line. If something isn't correct we clear the endpoint before returning with `retry == true`.

Comment (antonlisovenko) [reply pkg/broker/instance_operations.go:305]: ok, this makes sense 👍  - can we add the comment please?

**Review by antonlisovenko — APPROVED** [review]
Thanks for answering the questions - don't see anything suspicious considering my expertise in AOSB. 
Great work 👍

**Pull request merged**